src/compiler/program.ts

```diff
@@ -99,20 +106,27 @@ namespace ts {
         const directoryExists = !onlyRecordFailures && directoryProbablyExists(candidate, host);
         if (directoryExists && host.fileExists(packageJsonPath)) {
 
-            let jsonContent: { typings?: string };
+            let jsonContent: { typings?: string; main?: string };
 
             try {
                 const jsonText = host.readFile(packageJsonPath);
-                jsonContent = jsonText ? <{ typings?: string }>JSON.parse(jsonText) : { typings: undefined };
+                jsonContent = jsonText ? <{ typings?: string; main?: string }>JSON.parse(jsonText) : { typings: undefined };
             }
             catch (e) {
-                // gracefully handle if readFile fails or returns not JSON 
+                // gracefully handle if readFile fails or returns not JSON
                 jsonContent = { typings: undefined };
             }
 
             if (typeof jsonContent.typings === "string") {
                 const path = normalizePath(combinePaths(candidate, jsonContent.typings));
-                const result = loadNodeModuleFromFile(extensions, path, failedLookupLocation, !directoryProbablyExists(getDirectoryPath(path), host), host);
+                const result = loadNodeModuleFromFile(/*don't add extension*/[""], path, failedLookupLocation, !directoryProbablyExists(getDirectoryPath(path), host), host);
+                if (result) {
+                    return result;
+                }
+            }
+            if (typeof jsonContent.main === "string") {

```

Comment:
Not sure I get your comment or the `else if` prefix. Are you saying only try to load the file referenced by `main` if there was no `typings` setting?

I do notice one change I should make on review though: I should still use the `supportedExtensions` on the base path (i.e. if `main` points to `file1.js`, still look for `file1.d.ts` and `file1.ts` first (and also don't look for and load `.js` if not in the supportedExtensions list)). I'll update for that. Please clarify if I missed your original point.